datapath: masquerade IPv6 traffic when using iptables masquerading mode

fristonio opened the pull request:
See commit message for more detail.

This is with respect to an ongoing effort to improve support for DualStack K8s with Cilium and CI infrastructure.
Two new flags are introduced - `--enable-ipv6-masquerade` and `--enable-ipv4-masquerade` and the output of `cilium status` reflect the masquerading behavior enabled for the cluster.

```
$ cilium status
...
Host Routing:           Legacy
Masquerading:           IPTables [IPv4: Enabled, IPv6: Enabled]
...
```

Fixes https://github.com/cilium/cilium/issues/6320.

```release-note
Add the ability to masquerade IPv6 traffic when using iptables masquerading mode. This behavior can be enabled/disabled by using `enable-ipv6-masquerade` agent option.
```

christarazi commented on api/v1/models/masquerading.go:
```diff
@@ -34,8 +34,12 @@ type Masquerading struct {
-	// Any packet sent to IP addr belonging to CIDR will not be SNAT'd
-	SnatExclusionCidr string `json:"snat-exclusion-cidr,omitempty"`
+	// Any packet sent to IPv4 addr belonging to CIDR will not be SNAT'd
+	SnatExclusionCidrV4 string `json:"snat-exclusion-cidr-v4,omitempty"`
```
Nit: `SNATExclusionCIDRv4`

```
SNATExclusionCIDRv4 exempts SNAT from being performed on any packet sent to an IPv4 address that belongs inside this CIDR.
```

fristonio replied:
This is an autogenerated field using openapi spec. I don't think it's possible for us to change the name of the variable right? :confused:

ti-mo replied:
Yes, this whole file should be autogenerated and should not be modified manually. :)

christarazi commented on api/v1/models/masquerading.go:
```diff
@@ -34,8 +34,12 @@ type Masquerading struct {
+
+	// Any packet sent to IPv6 address belonging to this CIDR will not be SNAT'd.
+	// For IPv6 we only do masquerading in iptables mode.
+	SnatExclusionCidrV6 string `json:"snat-exclusion-cidr-v6,omitempty"`
```
Nit: similar naming convention and rewording as above. It's fine to leave `For IPv6 we only do masquerading in iptables mode.` as well.

christarazi reviewed (approved):
LGTM. I did not review the iptables rules, but have a minor nits for doc readability of the new SNAT exclusion arg.

brb commented on pkg/datapath/config.go:
```diff
@@ -123,14 +123,21 @@ type ConfigWriter interface {
+// packet sent from a local endpoint to an IP address belonging to the CIDR
+// should not be SNAT'd.
+func RemoteSNATDstAddrExclusionCIDRv6() *cidr.CIDR {
+	return node.GetIPv6AllocRange()
```
Is this pod CIDR node-local or cluster-wide?

fristonio replied:
This is the node-local IPv6 allocation range, similar to IPv4 alloc range.

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -828,6 +829,160 @@ func (m *IptablesManager) installForwardChainRules(ifName, localDeliveryInterfac
 	return nil
 }
 
+func (m *IptablesManager) installMasqueradeRules(prog, ifName, localDeliveryInterface,
```
Just a note that it would have been easier to review if you created the method in a separate commit.

fristonio replied:
Sorry if I misunderstood, but are you suggesting that in the first commit I add these rules wrapped inside an `if option.Config.EnableIPv6 { ... }` inside `InstallRules` and then in separate commit merge rules for IPv4 and IPv6 using a single function?

brb replied:
I meant in the first commit to move the existing IPv4 code into this method, and then in the second commit to add the IPv6 handling.

fristonio replied:
Makes sense, Done!

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -828,6 +829,160 @@ func (m *IptablesManager) installForwardChainRules(ifName, localDeliveryInterfac
 
+func (m *IptablesManager) installMasqueradeRules(prog, ifName, localDeliveryInterface,
+	snatDstExclusionCIDR, allocRange, hostMasqueradeIP string) error {
+	if option.Config.Masquerade && !option.Config.EnableBPFMasquerade {
```
While we are here, to reduce nestedness, you could `if option.Config.Masquerade && option.Config.EnableBPFMasquerade { return nil }`.

fristonio replied:
Makes sense, will do that.

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -929,147 +1084,39 @@ func (m *IptablesManager) InstallRules(ifName string) error {
-					return err
-				}
-			}
+		if err := m.installMasqueradeRules("iptables", ifName, localDeliveryInterface,
```
To make both cases symmetric, maybe we should wrap the invocation into `if option.Config.EnableIPv4 { ... }`?

fristonio replied:
This already seems to be wrapped inside if `option.Config.EnableIPv4{ ... }`
https://github.com/cilium/cilium/pull/14124/files#diff-465a6b3f4ab9e1e94fe39e5aefe81c41eaa1dc99c66500161072e564029c2f51R1052

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -929,147 +1084,39 @@ func (m *IptablesManager) InstallRules(ifName string) error {
-			}
+	if option.Config.EnableIPv6 {
+		// See comment above.
+		matchFromProxy := fmt.Sprintf("%#08x/%#08x", linux_defaults.MagicMarkIsProxy, linux_defaults.MagicMarkProxyMask)
```
Maybe we could create a function to install the MARK rule, and then call with with `"ip{,6}tables"` params?

brb reviewed (changes requested):
Thanks, a few nits :-)

Alternative solution to solve the asymmetric routing issue would be to force `pod -> remote node` traffic to be sent over the tunnel (https://github.com/cilium/cilium/issues/12624). However, considering that #6320 is quite requested feature, adding IPv6 masq makes sense.

In addition to the upgrade guide, we should update https://docs.cilium.io/en/v1.9/concepts/networking/masquerading/ too.

Next, is it possible in our CI to have a test for IPv6 masquerading?

Finally, do you plan to work on BPF-based masq (and ip-masq-agent) for IPv6?

fristonio committed:
docs: update masquerading docs with ipv6 traffic behaviour

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
test: add tests for iptables based masquerading for IPv6 traffic

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
iptables: refactor out host traffic mark rule to a separate method

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
iptables: refactor out ipv4 masquerading rules to a separate method

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
datapath: masquerade IPv6 traffic when using iptables masquerading mode

* This commits enables masquerading for IPv6 traffic using iptables when
  running in iptables masquerading mode.
* This resolves asymmetric traffic flows for NodePort services when
  running Cilium with IPv6 enabled. Earlier for IPv6 if a client on
  Node1 sends a request to Node2 NodePort service then the traffic is
  asymmetrically routed. The outgoing packets from Node1 are sent via host
  interface but the reply packets are sent back from the tunnel since
  there is no masquerading for IPv6 traffic.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

brb reviewed (approved):
Nice!

fristonio committed:
daemon: add separate flags to enable masquerading for ipv4 and ipv6 traffic

* This commit adds a two new flag `enable-ipv6-masquerade` and
  `enable-ipv4-masquerade` to enable masquerading for ipv6 and ipv4
  traffic respectivily.

* Deamon flag `--masquerade` is now deprecated in favour of the new
  `--enable-ipv4-masquerade` flag. The flag will be removed in 1.11
  cycle.

* Enabling IPv6 masquerading and BPF masquerading simulatneously will
  result in no masquerading as BPF masq is currently not supported for
  IPv6 traffic.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

gandro reviewed (approved):
Changes to health LGTM!

borkmann reviewed (approved):
LGTM, great work! Maybe upgrade guide optionally needs a deprecation notice and present the two new `enable-ipv{4,6}-masquerade` flags in the doc? We should also follow-up with an issue for implementing BPF-based masq for v6 as otherwise we won't be able to use the fast redirect datapath if users have v4+v6 enabled. Not blocking this one, of course, but something to resolve for 1.10 still.

fristonio committed:
docs: add deprecated and new options for masquerading to upgrade guide

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

joestringer commented on Documentation/operations/upgrade.rst:
```diff
@@ -337,12 +337,6 @@ Annotations:
-
-* ``k8s-force-json-patch``: This option does not have any effect for
-  environments running Kubernetes >= 1.13. Marked for removal in Cilium v1.11.
-
```
Why is this being removed? It looks like this is intended to be marked deprecated in 1.10 and removed in 1.11.

fristonio replied:
I think this has been answered in the below comment.

joestringer commented on Documentation/operations/upgrade.rst:
```diff
@@ -378,6 +381,12 @@ Deprecated Options
 
 * ``bpf-compile-debug``: This option does not have any effect since 1.10
   and is planned to be removed in 1.11.
+* ``k8s-force-json-patch``: This option does not have any effect for
+  environments running Kubernetes >= 1.13. Marked for removal in Cilium v1.11.
+* ``masquerade``: With the introduction of IPv6 masquerading this option has
+  been deprecated in favor of ``enable-ipv4-masquerade`` and is planned to
+  be removed in 1.11. For 1.10 release this option will have the same effect as
+  ``enable-ipv4-masquerade`` where both options must not be used simultaneously.
```
Ah, I see. I guess the question is whether we should have a separate "deprecated" vs. "removed" section. I don't know if we've had a discussion somewhere about this but it seems reasonable to me. :+1:

fristonio replied:
Yeah, there was one item each under both sections even though they had the same behaviour(deprecated in upcoming and removed in future releases). So I removed `Deprecated Option` section.

joestringer reviewed (changes requested):
This setting changes the commandline arguments, but does not update the Helm charts (`install/kubernetes/*`). Do you intend that the helm charts would also convert to the new split ipv4/ipv6 configuration, or that we retain the helm option for general `masquerade` and auto-enable both options?

Strictly speaking I think that it will work to not change `install/kubernetes` at this time, but we will need to revisit this before we properly remove the option and it'll be easy to forget if we don't do it now.

joestringer commented on cilium-health/launch/endpoint.go:
```diff
@@ -338,15 +341,14 @@ func LaunchAsEndpoint(baseCtx context.Context,
 		if err := routingConfig.Configure(healthIP,
-			mtuConfig.GetDeviceMTU(),
-			option.Config.Masquerade); err != nil {
+			mtuConfig.GetDeviceMTU(), masq); err != nil {
```
I don't see any IPv6 support in `routingConfig.Configure(...)` so maybe not worth making any changes here for now?

christarazi replied:
Correct, ENI mode does not support IPv6.

joestringer commented on pkg/datapath/maps/map.go:
```diff
@@ -86,7 +86,8 @@ func (ms *MapSweeper) deleteMapIfStale(path string, filename string, endpointID
 	globalCTinUse := ms.HasGlobalCT() || option.Config.EnableNodePort ||
-		!option.Config.InstallIptRules && option.Config.Masquerade
+		!option.Config.InstallIptRules && option.Config.EnableIPv4Masquerade &&
+			option.Config.EnableIPv6Masquerade
```
Should this be...?

```suggestion
		!option.Config.InstallIptRules && (option.Config.EnableIPv4Masquerade ||
			option.Config.EnableIPv6Masquerade)
```

fristonio replied:
Yeah, you are correct. Thanks for pointing this out.

joestringer reviewed (changes requested):
Couple of other nits from my codeowners (looked at all but the core iptables functionality which it looked like was already well-covered in other reviews)

fristonio committed:
helm: add helm option to configure masquerading behaviour for ipv{4,6} traffic

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

joestringer reviewed (commented):
Maybe also worthwhile to do a `git grep masquerade Documentation/` to check for any remaining references to the old options. At a glance I see a few in the chaining modes, we could either expand those configurations for both of the new options or consider having the existing option act as a single "override".
